src/learning/agglomerative.rs

```diff
@@ -0,0 +1,415 @@
+    Ward,
+    /// Ward I,
+    /// See "Ward’s Hierarchical Agglomerative Clustering Method:
+    /// Which Algorithms Implement Ward’s Criterion? (Murtagh, 2014)"
+    Ward1,
+    /// Ward II
+    Ward2,
+}
+
+impl Metrics {
+
+    // calculate distance using Lance-Williams algorithm
+    fn dist(&self, ci: &Cluster, cj: &Cluster, ck: &Cluster, dmat: &DistanceMatrix) -> f64 {
+
+        let dik = dmat.get(ck.id, ci.id);
+        let djk = dmat.get(ck.id, cj.id);
+
+        match self {
+            &Metrics::Single => {
+                // 0.5 * dik + 0.5 * djk + 0. * dij - 0.5 * (dik - djk).abs()
+                dik.min(djk)
+            },
+            &Metrics::Complete =>  {
+                // 0.5 * dik + 0.5 * djk + 0. * dij + 0.5 * (dik - djk).abs()
+                dik.max(djk)
+            },
+            &Metrics::Average =>  {
+                let s = ci.size + cj.size;
+                ci.size / s * dik + cj.size / s * djk
+            },
+            &Metrics::Centroid => {
+                let s = ci.size + cj.size;
+                let ai = ci.size / s;
+                let aj = cj.size / s;
+                let dij = dmat.get(ci.id, cj.id);
+                ai * dik + aj * djk - ai * aj * dij
+            },
+            &Metrics::Median => {
+                let dij = dmat.get(ci.id, cj.id);
+                0.5 * dik + 0.5 * djk - 0.25 * dij
+            },
+            &Metrics::Ward1  => {
+                let s = ci.size + cj.size + ck.size;
+                let dij = dmat.get(ci.id, cj.id);
+                (ci.size + ck.size) / s * dik + (cj.size + ck.size) / s * djk - ck.size / s * dij
+            },
+            &Metrics::Ward | &Metrics::Ward2 => {
+                let s = ci.size + cj.size + ck.size;
+                let dij = dmat.get(ci.id, cj.id);
+                ((ci.size + ck.size) / s * dik.powf(2.) + (cj.size + ck.size) / s * djk.powf(2.) - ck.size / s * dij.powf(2.)).sqrt()
+            }
+        }
+    }
+}
+
+struct Cluster {
+    /// Cluster id
+    id: usize,
+    /// Number of nodes (rows) which belongs to cluster
+    /// to avoid cast in the algorithm, store it as f64
+    size: f64,
+    /// Row ids belong to the cluster
+    nodes: Vec<usize>,
+}
+
+impl Cluster {
+
+    /// Create new cluster
+    fn new(id: usize, nodes: Vec<usize>) -> Cluster {
+        Cluster {
+            id: id,
+            size: 1.,
+            nodes: nodes,
+        }
+    }
+
+    /// Create new cluster merging left and right
+    fn from_clusters(id: usize, left: Cluster, mut right: Cluster) -> Cluster {
+        let mut new_nodes = left.nodes;
+        new_nodes.append(&mut right.nodes);
+        Cluster {
+            id: id,
+            size: left.size + right.size,
+            nodes: new_nodes
+        }
+    }
+}
+
+/// Distance Matrix
+#[derive(Debug)]
+struct DistanceMatrix {
+    // Distance is symmetric, no need to hold all pairs
+    // use HashMap to easier update
+    data: HashMap<(usize, usize), f64>
```

Comment:
This is for simple implementation rather than memory efficiency. 

Because distance matrix should keep only remaining nodes and merged clusters, mapping node id to matrix cell (of triangular flattened to vector) makes impl complex. 

I think it is OK ATM as hierarchical clustering is not likely to be used for larger data, but appreciated if anyone lmk better impl.